proposals/accepted/50-ide-test-integration.md

```diff
@@ -0,0 +1,219 @@
+
+-- | A lazy list in order to support arbitrary execution
+-- order and streaming between the test runner and IDE
+data TestResults = TestResults [TestCaseResult]
+
+data TestCaseResult = TestCaseResult {
+    -- | Represents the path to get to this test case from the
+    -- TestTree, where each element represents a node in the original
+    -- TestTree.
+    path :: NonEmpty Text,
+    result :: TestResult,
+    timeTaken :: Maybe Double
+  }
+
+data TestResult = Success | Skipped | Failure FailureReason
+
+newtype FailureReason = FailureReason Text
+```
+
+It will also contain the logic to encode this datatype
+to JSON, for consumption by IDEs.
+
+------------
+
+For testing frameworks that want to support IDE integration,
+they should provide support for the following commands:
+
+```bash
+./test-suite test-tree --output="FILE_NAME"
+./test-suite run-tests --input="FILE" --output="FILE"
+
+# Sample usages:
+> cabal test test-suite-name \
+      \--test-options='test-tree --output="output.tmp.json"'
+> cabal test test-suite-name \
+      \--test-options='run-tests --input="input.tmp.json" --output="output.tmp.json"'
+```
+
+The `test-tree` command outputs a JSON-serialized `TestTree` to the given
+`output` file.
+
+The `run-tests` command reads a JSON-serialized `RunTestsRequest` from
+`input`, and, after executing tests, writes the result to the `output` file.
+
+`hs-test` will provide helper functions for these serialization tasks.
+
+-------------
+
+For IDEs, they will utilize these commands to:
+ -  find the test tree
+ -  execute a test run request, and display the results to the user.
+
+One problem that could occur is that having the IDE run `cabal test` to
+reindex the `TestTree` is too slow and incurs too large a latency for
+the user. However, this problem is no worse than the current situation,
+in which the user has to call `cabal test` anyway in order to run the tests.
+
+Furthermore, IDEs often having access to stale artifacts for code that used
+to compile but no longer compiles, therefore, the IDE could run the `test-tree`
+command while still displaying a mostly-correct test tree and inline hints
+in the user's IDE. Once the call to `test-tree` has completed, the new test
+tree can be displayed to the user.
+
+## Alternative Approaches
+
+### A typeclass driven approach with HLS integration
+
+Described in the following [discourse comment](https://discourse.haskell.org/t/hf-coordination-for-test-framework-ide-integration/5221/2?u=santiweight)
+
+In summary, take each function signature and detect whether said function
+is runnable as a test based on whether its type is a member of the
+proposed typeclass `IsTest`. Testing frameworks will each provide their
+own `IsTest` instances for their common testing types.
+
+I have chosen the option in this proposal for the following reasons:
+ 1. Such an approach would run Haskell code in GHCi, which is slower, and
+ would therefore scale poorly for production codebases.
```

Comment:
I don't think this is actually true - the type class based approach seems to me like it could support compiled or interpreted mode, and that GHCi is an implementation strategy not forced by the design. Is there something I'm missing?

Reply:
I also don't understand this. For running the tests you could still run them through cabal or whatever, assuming that you have some way of knowing how to do that.